src/pilot/dell_nfv.py

```diff
@@ -0,0 +1,353 @@
+                    vcpu_pin_set +
+                    "\"|' " +
+                    file_path)
+
+            for cmd in cmds:
+                status = os.system(cmd)
+                logger.debug("cmd: {}".format(cmd))
+            if status != 0:
+                raise Exception(
+                    "Failed to execute the command {}"
+                    " with error code {}".format(
+                        cmd, status))
+
+        except Exception as error:
+            message = "Exception {}: {}".format(
+                type(error).__name__, str(error))
+            logger.error("{}".format(message))
+            raise Exception(
+                "Failed to modify the dell_environment.yaml"
+                " at location {}".format(file_path))
+
+    def get_dell_compute_nodes_hostnames(self, nova):
+        try:
+            logger.info("Getting dellnfv compute node hostnames")
+            # Create host object
+            host_obj = hosts.HostManager(nova)
+
+            # Get list of dell nfv nodes
+            dell_hosts = []
+
+            for host in host_obj.list():
+                if "dell-compute" in host.host_name:
+                    hostname = str(host.host_name)
+                    dell_hosts.append(hostname)
+            return dell_hosts
+        except Exception as error:
+            message = "Exception {}: {}".format(
+                type(error).__name__, str(error))
+            logger.error("{}".format(message))
+            raise Exception("Failed to get the Dell Compute nodes.")
+
+    def edit_aggregate_environment_file(
+            self, hostname_list):
+        logger.info("Editing create aggregate environment file")
+        file_path = home_dir \
+            + '/pilot/templates/create_aggregate_environment.yaml'
+        if not os.path.isfile(file_path):
+            raise Exception(
+                "The create_aggregate_environment.yaml file does not exist")
+        cmd = (
+            'sed -i "s|hosts:.*|hosts: ' +
+            str(hostname_list) +
+            '|" ' +
+            file_path)
+
+        status = os.system(cmd)
+        logger.info("cmd: {}".format(cmd))
+        if status != 0:
+            raise Exception(
+                "Failed to execute the command {}"
+                " with error code {}".format(
+                    cmd, status))
+
+    def create_aggregate(self):
+        UC_AUTH_URL, UC_PROJECT_ID, UC_USERNAME, UC_PASSWORD = \
+            CredentialHelper.get_overcloud_creds()
+        # Create nova client object
+        nova = nvclient.Client(
+            2,
+            UC_USERNAME,
+            UC_PASSWORD,
+            UC_PROJECT_ID,
+            UC_AUTH_URL)
+        hostname_list = self.get_dell_compute_nodes_hostnames(nova)
+        self.edit_aggregate_environment_file(
+            hostname_list)
+        env_opts = \
+            " -e ~/pilot/templates/create_aggregate_environment.yaml"
+
+        cmd = self.overcloudrc + "openstack stack create " \
```

Comment:
Rather than using the heat client for creating the stack, i think it would be better to use the heat python client, and let the method heat_client.stacks.create do the job for us. Sourcing the overcloudrc file would become useless as a result.
It can be achieved considering the following (take it as an example):
    os_auth_url, os_tenant_name, os_username, os_password = \
        CredentialHelper.get_overcloud_creds()

    keystone_client = clients.get_keystone_client(os_username,
                                                  os_password,
                                                  os_tenant_name,
                                                  os_auth_url)

    heat_url = keystone_client.service_catalog.url_for(
        service_type='orchestration',
        endpoint_type='publicURL')

    heat_client = HeatClient(endpoint=heat_url,
                             token=keystone_client.auth_token)

    template = open(template)
    environment = open(environment)
    stack = heat_client.stacks.create(stack_name=stack_name,
                                      template=template.read(),
                                      environment=environment.read())
    template.close()
    environment.close()

Reply:
Done
Take a look